pkg/pgmodel/labels.go

```diff
@@ -23,53 +24,56 @@ type Labels struct {
 	str        string
 }
 
-// EmptyLables returns an empty Labels object
-func EmptyLables() Labels {
-	return Labels{}
-}
+var LabelsInterner = sync.Map{}
 
-// LabelsFromSlice converts a labels.Labels to a Labels object
-func LabelsFromSlice(ls labels.Labels) (Labels, error) {
-	length := len(ls)
-	labels := Labels{
-		names:  make([]string, length),
-		values: make([]string, length),
+func GetLabels(str string) (l *Labels) {
+	val, ok := LabelsInterner.Load(str)
+	if !ok {
+		return
 	}
+	l = val.(*Labels)
+	return
+}
 
-	labels.metricName = ""
-	for i, l := range ls {
-		labels.names[i] = l.Name
-		labels.values[i] = l.Value
-		if l.Name == MetricNameLabelName {
-			labels.metricName = l.Value
-		}
-	}
+func SetLabels(str string, lset *Labels) *Labels {
+	val, _ := LabelsInterner.LoadOrStore(str, lset)
+	return val.(*Labels)
+}
 
-	err := initLabels(&labels)
-	return labels, err
+// LabelsFromSlice converts a labels.Labels to a Labels object
+func LabelsFromSlice(ls labels.Labels) (*Labels, error) {
+	ll := make([]prompb.Label, len(ls))
+	for i := range ls {
+		ll[i].Name = ls[i].Name
+		ll[i].Value = ls[i].Value
+	}
+	l, _, err := labelProtosToLabels(ll)
+	return l, err
 }
 
 // initLabels intializes labels
-func initLabels(l *Labels) error {
-
-	if !sort.IsSorted(l) {
-		sort.Sort(l)
+func getStr(labels []prompb.Label) (string, error) {
+
+	if !sort.SliceIsSorted(labels, func(i, j int) bool {
```

Comment:
nit: can we format this  in a saner way. Also these functions are identical, just make a function variable?